### DISCUSSION TITLE
Is there  a way to enable filtering and pagination for datatables

### DISCUSSION THREAD

**Discussion by gusmb** [squidfunk/mkdocs-material#5011] [Q&A] [answered]
I am rendering markdown documentation from templates with large datatables. Right now I have enabled the sortable columns feature as explained in the docs. Is there a way to enable filetering on columns and pagination to avoid printing the whole table at once?

**Comment by squidfunk** [comment, marked as answer]
Material for MkDocs doesn't provide this functionality out of the box, but it's not difficult to add by using a third-party library, for example something like [TableFilter](https://www.tablefilter.com/). Integration is similar to [TableSort](https://squidfunk.github.io/mkdocs-material/reference/data-tables/#sortable-tables).

**Reply by gusmb** [reply]
I was playing around with dataTables and got what I was looking for. The downside is that they are not markdown tables but plain HTML syntax. Wish there was a native markdown solution at some point in the future :)

**Reply by squidfunk** [reply]
Yeah, Markdown tables are meh. I think I've seen some extensions to the standard Markdown table syntax, but can't remember.

**Comment by plbremer** [comment]
@gusmb @squidfunk 
Are you able to advise? I am trying to do what Gus did for my own documents.

I have:
1) Updated the extra_javascript portion of mkdocs.yaml to remove conflicts from filter
2) Put what I believe to be the right unpkg link
```
extra_javascript:
  # - https://unpkg.com/tablesort@5.3.0/dist/tablesort.min.js
  # - javascripts/tablesort.js
  - https://unpkg.com/tablefilter@0.7.3/dist/tablefilter/tablefilter.js
  - javascripts/tablefilter.js
```

3) Attempted to instantiate a filter class in a way analogous to the way presented in the Tablesort example
```
document$.subscribe(function() {
//   var myTF = new TableFilter(document.querySelector(".my-table"));
  var tables = document.querySelectorAll("article table:not([class])")
  tables.forEach(function(table) {
    new TableFilter(table)
  })
})
```

Of course, "nothing shows up". I suspect that my .js gist is not working. Any advice? Thanks

**Reply by kamilkrzyskow** [reply]
~~Discussion moved to https://github.com/squidfunk/mkdocs-material/discussions/6597~~ 

Edit by @squidfunk: Closed as duplicate. Please comment here, as long as you're discussing the exact same problem.

**Reply by plbremer** [reply]
Here I provide MWE - 

What I have, and what is not working, is [9.5.3-plbremer_filter.zip](https://github.com/squidfunk/mkdocs-material/files/13837014/9.5.3-plbremer_filter.zip).

I was hoping that if I closely paralled the instructions given in https://squidfunk.github.io/mkdocs-material/reference/data-tables/#sortable-tables , as shown in [9.5.3-plbremer_sort.zip](https://github.com/squidfunk/mkdocs-material/files/13837016/9.5.3-plbremer_sort.zip), that I would be successful.

That is not the case.

**Reply by alexvoss** [reply]
This JavaScript works for me:

```javascript
document$.subscribe(function() {
  var tables = document.querySelectorAll("article table:not([class])")
  tables.forEach(function(table) {
    var tf = new TableFilter(table, {
      base_path: "https://unpkg.com/tablefilter@0.7.3/dist/tablefilter/"
    })  
    tf.init()
  })  
})
```

This is just following the [tablefilter documentation](https://www.tablefilter.com) (see Usage section).

**Reply by plbremer** [reply]
Thanks - I didn't realize that it worked like that (ie anything in script tags would work in this context)

**Reply by revati-naik** [reply]
Hi @alexvoss,

I am using your example of the `.js` script. However, it isn't working out for me. I am in doubt about the `base_path`. Is that what it should be?

**Reply by plbremer** [reply]
@revati-naik 
fyi, this works for me
https://github.com/linkml/linkml-project-cookiecutter/pull/104/files#diff-da819ae4cffe1cf59b69392096d242f366db550737272565e652a0f207c27119

**Reply by alexvoss** [reply]
@plbremer, thanks for confirming that it works as I don't use it myself.

@revati-naik do you have the lines in the `mkdocs.yml` that load both the tablefilter code and the extra code I posted to initialize it?

```
extra_javascript:
  - https://unpkg.com/tablefilter@0.7.3/dist/tablefilter/tablefilter.js
  - assets/javascript/tablefilter.js
```

The latter you'd need to adapt to wherever you put that code.

 If yes, feel free to post a [minimal reproduction](https://squidfunk.github.io/mkdocs-material/guides/creating-a-reproduction/) and I will look at it.

**Reply by revati-naik** [reply]
@alexvoss Here you go

tablefilter.js
```
document$.subscribe(function() {
    var tables = document.querySelectorAll("article table:not([class])")
    tables.forEach(function(table) {
      var tf = new TableFilter(table, {
        base_path: "https://unpkg.com/tablefilter@0.7.3/dist/tablefilter/",
        filters_row_index: 0,
      })  
      tf.init()
    })  
  })
  ```
  
mkdocs.yml
  ```
  extra_javascript:
  - https://unpkg.com/tablefilter@0.7.3/dist/tablefilter/tablefilter.js
  - javascript/tablefilter.js
  ```
  
  EDIT: Great! I got this working. Adding this really messes up my existing tables :D
  
  What I am looking for is adding pagination and search in table features. Is this possible? I am looking at the [filter configuration page](https://github.com/koalyptus/TableFilter/wiki/1.0-Configuration), but not something I see directly in this list. Is there some way to get this feature integrated?

**Reply by alexvoss** [reply]
Please provide a [minimal reproduction](https://squidfunk.github.io/mkdocs-material/guides/creating-a-reproduction/) so I can look at the problem. I also need to know *what* is not working. Two code snippets don't tell me that. What is it that this should be doing or not doing and are you getting any messages on the JavaScript console?

**Reply by plbremer** [reply]
it would be great if you could post how you got it working for future viewers

**Reply by revati-naik** [reply]
I had to mention the right `base_path` as suspected above. Fixing that, made the filter available. Rest everything works as expected. Thanks!

**Reply by mihaigalos** [reply]
I mean the functionality works but is ugly to say the least:
<img width="717" alt="grafik" src="https://github.com/squidfunk/mkdocs-material/assets/16443090/aba1d6ef-ea21-4d61-bccf-aeb63b53d90e">

Any way to just add a small filter icon and only then show the input fields for the filter? Something like this:
<img width="1470" alt="grafik" src="https://github.com/squidfunk/mkdocs-material/assets/16443090/d702c739-8b69-4c4d-9e34-465fca4c25f7">

**Reply by alexvoss** [reply]
You're going to have to ask the people who wrote [tablefilter](https://github.com/koalyptus/TableFilter).

**Reply by mihaigalos** [reply]
You're right. Sorry, last commit 4y ago. I thought this thread would be more appropriate. 😅

**Reply by mihaigalos** [reply]
Anyway, for future reference, this is as far as I've gotten:

<img width="375" alt="grafik" src="https://github.com/squidfunk/mkdocs-material/assets/16443090/d80bef59-2f73-4178-8e8d-22df61772e7c">


<details>
  <summary>Code - click to preview!</summary>

`index.html`:

```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Filterable Table</title>
  <link rel="stylesheet" href="styles.css">
</head>
<body>
  <article>
    <table id="myTable">
      <thead>
        <tr>
          <th>Name</th>
          <th>Age</th>
          <th>City</th>
        </tr>
      </thead>
      <tbody>
        <tr><td>Alice</td><td>24</td><td>New York</td></tr>
        <tr><td>Bob</td><td>30</td><td>Los Angeles</td></tr>
        <tr><td>Charlie</td><td>22</td><td>Chicago</td></tr>
        <!-- Add more rows as needed -->
      </tbody>
    </table>
  </article>
  <script src="script.js"></script>
</body>
</html>

```
`script.js`:
```js
document.addEventListener("DOMContentLoaded", function() {
  var table = document.querySelector('table');
  var headers = table.querySelectorAll("th");

  headers.forEach(function(header, index) {
    // Create a container for the header content
    var container = document.createElement('div');
    container.className = 'header-container';
    container.innerHTML = header.innerHTML;

    // Create filter icon
    var filterIcon = document.createElement('span');
    filterIcon.className = 'filter-icon';
    filterIcon.innerHTML = '🌪️';

    // Create clear icon
    var clearIcon = document.createElement('span');
    clearIcon.className = 'clear-icon';
    clearIcon.innerHTML = '🧹';

    // Create filter input
    var input = document.createElement('input');
    input.type = 'text';
    input.className = 'filter-input';

    // Append icons and input to the container
    container.appendChild(filterIcon);
    container.appendChild(clearIcon);
    container.appendChild(input);

    // Replace header content with the container
    header.innerHTML = '';
    header.appendChild(container);

    filterIcon.addEventListener("click", function(event) {
      event.stopPropagation();
      hideAllInputs();
      if (input.style.display === "none" || input.style.display === "") {
        input.style.display = "block";
        input.style.left = filterIcon.offsetLeft + "px";
        input.style.top = filterIcon.offsetTop + filterIcon.offsetHeight + "px";
      } else {
        input.style.display = "none";
      }
      input.focus();
    });

    clearIcon.addEventListener("click", function(event) {
      event.stopPropagation();
      input.value = "";
      filterTable(table, index, "");
      input.style.display = "none";
      clearIcon.style.display = "none";
    });

    input.addEventListener("input", function() {
      filterTable(table, index, input.value);
      clearIcon.style.display = input.value ? "inline" : "none";
    });

    document.addEventListener("click", function(event) {
      if (input.style.display === "block" && !input.contains(event.target) && !filterIcon.contains(event.target)) {
        input.style.display = "none";
      }
    });

    input.addEventListener("click", function(event) {
      event.stopPropagation();
    });
  });

  function hideAllInputs() {
    var inputs = document.querySelectorAll(".filter-input");
    inputs.forEach(function(input) {
      input.style.display = "none";
    });
  }

  function filterTable(table, column, query) {
    var rows = table.querySelectorAll("tbody tr");

    rows.forEach(function(row) {
      var cell = row.cells[column].textContent.trim();
      if (cell.toLowerCase().includes(query.toLowerCase())) {
        row.style.display = "";
      } else {
        row.style.display = "none";
      }
    });
  }
});
```
`styles.css`:
```css
.filter-icon, .clear-icon {
  cursor: pointer;
  margin-left: 5px;
}
.filter-input {
  display: none;
  position: absolute;
  background: white;
  border: 1px solid #ccc;
  z-index: 1000;
  padding: 5px;
}
.header-container {
  position: relative;
}
.clear-icon {
  display: none;
}
```
</details>
